metrics-operator/controllers/analysis/controller.go

```diff
@@ -0,0 +1,160 @@
+
+	//find AnalysisDefinition to have the collection of Objectives
+	analysisDef := &metricsapi.AnalysisDefinition{}
+	if analysis.Spec.AnalysisDefinition.Namespace == "" {
+		analysis.Spec.AnalysisDefinition.Namespace = a.Namespace
+	}
+	err := a.Client.Get(ctx,
+		types.NamespacedName{
+			Name:      analysis.Spec.AnalysisDefinition.Name,
+			Namespace: analysis.Spec.AnalysisDefinition.Namespace},
+		analysisDef,
+	)
+
+	if err != nil {
+		if errors.IsNotFound(err) {
+			a.Log.Info(err.Error() + ", ignoring error since object must be deleted")
+			return ctrl.Result{Requeue: true, RequeueAfter: 10 * time.Second}, nil
+		}
+		a.Log.Error(err, "Failed to retrieve the AnalysisDefinition")
+		return ctrl.Result{RequeueAfter: 10 * time.Second}, nil
+	}
+
+	tempAnalysisDef := analysisDef
+	var done map[string]metricsapi.ProviderResult
+	if analysis.Status.Raw != "" {
+		done = a.ExtractMissingObj(tempAnalysisDef, analysis.Status.StoredValues)
+	}
+
+	//create multiple workers handling the Objectives
+	childCtx, wp := a.NewWorkersPoolFactory(ctx, analysis, tempAnalysisDef, a.MaxWorkers, a.Client, a.Log, a.Namespace)
+
+	res, err := wp.DispatchAndCollect(childCtx)
+	if err != nil {
+		a.Log.Error(err, "Failed to Collect all SLOs, caching collected values")
+		analysis.Status.StoredValues = res
+		err = a.updateStatus(ctx, analysis)
+		return ctrl.Result{RequeueAfter: 10 * time.Second}, err
+	}
+
+	maps.Copy(res, done)
+
+	eval := a.Evaluate(res, analysisDef)
+	analysisResultJSON, err := json.Marshal(eval)
+	if err != nil {
+		a.Log.Error(err, "Could not marshal status")
+	}
+	analysis.Status.Raw = string(analysisResultJSON)
+	if eval.Warning {
+		analysis.Status.Warning = true
+	}
+	analysis.Status.Pass = eval.Pass
+	err = a.updateStatus(ctx, analysis)
+
+	return ctrl.Result{}, err
+}
+
+func (a *AnalysisReconciler) updateStatus(ctx context.Context, analysis *metricsapi.Analysis) error {
+	if err := a.Client.Status().Update(ctx, analysis); err != nil {
+		a.Log.Error(err, "Failed to update the Analysis status")
+		return err
+	}
+	return nil
+}
+
+// SetupWithManager sets up the controller with the Managea.
+func (a *AnalysisReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&metricsapi.Analysis{}, builder.WithPredicates(predicate.GenerationChangedPredicate{})).
+		Complete(a)
+}
+
+func (a *AnalysisReconciler) ExtractMissingObj(def *metricsapi.AnalysisDefinition, status map[string]metricsapi.ProviderResult) map[string]metricsapi.ProviderResult {
+	toDo := []metricsapi.Objective{}
+	done := make(map[string]metricsapi.ProviderResult, len(status))
+	for _, obj := range def.Spec.Objectives {
+		if value, ok := status[common.ComputeKey(obj.AnalysisValueTemplateRef)]; ok {
+			if value.Err != "" {
+				toDo = append(toDo, obj)
+			} else {
+				done[common.ComputeKey(obj.AnalysisValueTemplateRef)] = status[common.ComputeKey(obj.AnalysisValueTemplateRef)]
+			}
+		}
+	}
+	def.Spec.Objectives = toDo
```

Comment:
modifying the spec of the analysisDefinition in this function might have some unwanted side effects (e.g. it could be that we end up with an empty list of objectives here, which will then be passed to the `Evaluate` method later). I would propose to leave the AnalysisDefinition unmodified in this function and make it return a list of objectives instead